<a href="https://colab.research.google.com/github/Rehman-Cg7/AIML/blob/main/Task3.ipynbweek3_task1_model_evaluation_tuning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [176]:
# from google.colab import drive
# drive.mount('/content/drive')

In [177]:
# import pandas as pd
# import numpy as np

# df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Titaanic Dataset.csv')

# df.head()

In [178]:
# df.info()

In [179]:
# df.describe()

In [180]:
# df.isnull().sum()

In [181]:
# df.dtypes

In [182]:
# import pandas as pd
# import seaborn as sns
# import matplotlib.pyplot as plt
# import numpy as np

# url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
# titanic = pd.read_csv(url)

# titanic.head()
# titanic.isnull().sum()

## Handling Missing Values

- Age: Median se fill kiya kyunke age mein outliers hain, aur mean unse affect ho jata — median zyada stable hai.
- Cabin: Column drop kar diya kyunke isme 70%+ values missing thi — itna missing data fill karna galat estimate dega.
- Embarked: Sirf 2 missing values thi, is liye mode (sabse common value) se fill kiya.

In [183]:
# # Age ko median se fill karo
# titanic['Age'].fillna(titanic['Age'].median(), inplace=True)

# # Embarked ko mode se fill karo
# titanic['Embarked'].fillna(titanic['Embarked'].mode()[0], inplace=True)

# # Cabin column drop karo (bohot missing values hain)
# titanic.drop(columns=['Cabin'], inplace=True)

# # Dobara check karo
# titanic.isnull().sum()

#4 Visualizations

###1. Histogram — Age distribution

In [184]:
# plt.figure(figsize=(8,5))
# sns.histplot(titanic['Age'], kde=True, bins=30)
# plt.title('Age Distribution')
# plt.xlabel('Age')
# plt.show()

###2. Boxplot — Age by Survival

In [185]:
# plt.figure(figsize=(8,5))
# sns.boxplot(x='Survived', y='Age', data=titanic)
# plt.title('Age vs Survival')
# plt.show()

###3. Bar chart — Survival by Gender

In [186]:
# plt.figure(figsize=(8,5))
# sns.barplot(x='Sex', y='Survived', data=titanic)
# plt.title('Survival Rate by Gender')
# plt.ylabel('Survival Rate')
# plt.show()

###4. Correlation Heatmap

In [187]:
# plt.figure(figsize=(10,6))
# numeric_cols = titanic.select_dtypes(include=[np.number])
# sns.heatmap(numeric_cols.corr(), annot=True, cmap='coolwarm')
# plt.title('Correlation Heatmap')
# plt.show()

## Which feature affects survival most?

Sex (gender) survival ko sabse zyada affect karta hai. Bar chart mein clearly dikha
ke females ki survival rate males se kaafi zyada thi — "women and children first"
policy ke wajah se. Heatmap mein bhi Pclass (ticket class) ka strong correlation
survival ke saath dikha — higher class passengers ki survival chance zyada thi,
shayad unko lifeboats tak pehle access mila.

##Task#3
###Classification (Logistic Regression Algurithm)

In [188]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix
import pandas as pd

In [189]:
# Sex aur Embarked ko one-hot encode karo
titanic_encoded = pd.get_dummies(titanic, columns=['Sex', 'Embarked'], drop_first=True)

titanic_encoded.head()

,PassengerId,Survived,Pclass,Name,Age,SibSp,Parch,Ticket,Fare,Sex_male,Embarked_Q,Embarked_S
0,1,0,3,"Braund, Mr. Owen Harris",22.0,1,0,A/5 21171,7.2500,True,False,True
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1,0,PC 17599,71.2833,False,False,False
2,3,1,3,"Heikkinen, Miss. Laina",26.0,0,0,STON/O2. 3101282,7.9250,False,False,True
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,0,113803,53.1000,False,False,True
4,5,0,3,"Allen, Mr. William Henry",35.0,0,0,373450,8.0500,True,False,True


In [190]:
# Un columns ko drop karo jo useful nahi (Name, Ticket, PassengerId text hain, model use nahi kar sakta)
features = titanic_encoded.drop(columns=['Survived', 'Name', 'Ticket', 'PassengerId'])

X = features   # inputs — jinse prediction karni hai
y = titanic_encoded['Survived']   # output — jo predict karna hai

X.head()

,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
0,3,22.0,1,0,7.2500,True,False,True
1,1,38.0,1,0,71.2833,False,False,False
2,3,26.0,0,0,7.9250,False,False,True
3,1,35.0,1,0,53.1000,False,False,True
4,3,35.0,0,0,8.0500,True,False,True


In [191]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Training data size:", X_train.shape)
print("Testing data size:", X_test.shape)

Training data size: (712, 8)
Testing data size: (179, 8)


In [192]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [193]:
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy: {accuracy * 100:.2f}%")

Model Accuracy: 81.01%


In [194]:
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[90 15]
 [19 55]]


## Confusion Matrix Explanation

Confusion matrix se pata chalta hai ke model ne 179 test passengers mein se 145 ki
prediction sahi ki (90 True Negatives + 55 True Positives), jo model ki accuracy
81% ke barabar hai. Model ne 15 logon ko galat "survived" predict kiya (False
Positive), aur 19 logon ko galat "not survived" predict kiya (False Negative).
Ye dikhata hai ke model reasonably accha perform kar raha hai, lekin thora sa
False Negatives ki taraf lean karta hai — matlab kabhi kabhi asal survivors ko
miss kar deta hai.

#Task#3
Next Level

In [195]:
# from sklearn.datasets import fetch_california_housing
# import pandas as pd

# # Dataset load karo
# housing = fetch_california_housing(as_frame=True)
# df = housing.frame

# df.head()

In [196]:
# df.columns

In [197]:
# # Features jo humein lagta hai price affect karte hain
# features = ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms']

# X = df[features]
# y = df['MedHouseVal']

# X.head()

###I selected these four features:

* **MedInc (Median Income):** Areas with higher median income generally have more expensive houses.
* **HouseAge:** The age of a house can affect its value.
* **AveRooms:** Houses with more rooms are generally larger and may have higher prices.
* **AveBedrms:** The number of bedrooms is also related to the size and price of a house.


Train-Test **`Split`**

In [198]:
# from sklearn.model_selection import train_test_split

# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# print("Training size:", X_train.shape)
# print("Testing size:", X_test.shape)

# **Linear Regression Model Training**

In [199]:
# from sklearn.linear_model import LinearRegression

# model = LinearRegression()
# model.fit(X_train, y_train)

## Predictions Karo

In [200]:
# y_pred = model.predict(X_test)

# # Pehli 5 predictions dekho
# print("Predicted:", y_pred[:5])
# print("Actual:", y_test[:5].values)

# Model Evaluation — RMSE Aur R²

In [201]:
# from sklearn.metrics import mean_squared_error, r2_score
# import numpy as np

# rmse = np.sqrt(mean_squared_error(y_test, y_pred))
# r2 = r2_score(y_test, y_pred)

# print(f"RMSE: {rmse:.4f}")
# print(f"R² Score: {r2:.4f}")

**NOW**
# Scatter Plot — Predicted vs Actual

In [202]:
# import matplotlib.pyplot as plt

# plt.figure(figsize=(8,6))
# plt.scatter(y_test, y_pred, alpha=0.5)
# plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], 'r--', lw=2)
# plt.xlabel('Actual Price')
# plt.ylabel('Predicted Price')
# plt.title('Predicted vs Actual House Prices')
# plt.show()

**How to read this chart:** Each dot represents a house. The **X-axis** shows the actual price, while the **Y-axis** shows the model’s predicted price. The red dashed line represents the **perfect prediction line**—if the model were perfect, all the dots would lie exactly on this line. The closer the dots are to the line, the better the model’s predictions. If the dots are widely scattered, it indicates that the model has a higher prediction error.


## R² Score Explanation (Non-Technical)

My model’s **R² score is 0.5089**, which means that the model explains about **50.89% of the variation in house prices** using these four features: income, age, rooms, and bedrooms.

In simple terms, if we provide the model with information about a house’s income level, age, and number of rooms and bedrooms, it can explain around **50.89% of the variation in its price**. The remaining **49.11%** is influenced by other factors that are not included in our model, such as the exact location, school quality, crime rate, and other housing-related factors.

# **RMSE: 0.8022**

# **R² Score: 0.5089**


# Model Evaluation & Tuning: Beyond Accuracy

**Classification** Report

In [203]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.83      0.86      0.84       105
           1       0.79      0.74      0.76        74

    accuracy                           0.81       179
   macro avg       0.81      0.80      0.80       179
weighted avg       0.81      0.81      0.81       179



# **Step 2: Written Explanation — Why Accuracy Is Misleading (Markdown Cell)**
## Why Accuracy Can Be Misleading

If the dataset is imbalanced (e.g., 90% of the logs are in one class and 10% in another),
the model may always predict the majority class and still show high accuracy,
without actually learning anything. For example, if 90% of the logs are "not survived" and
the model always predicts "not survived," its accuracy will be 90%, but it will never
correctly identify any survivors. This is why Precision, Recall, and F1-score are important—they indicate how accurately the model is identifying the minority class (here "survived"), not just overall accuracy.

# Step 3: Hyperparameter Tuning — GridSearchCV


Now let's systematically test the model settings (hyperparameters) to find the best combination, rather than guessing.

In [204]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

# Jo parameters test karne hain, unki grid banao
param_grid = {
    'C': [0.01, 0.1, 1, 10, 100],
    'solver': ['liblinear', 'lbfgs']
}

grid_search = GridSearchCV(
    LogisticRegression(max_iter=1000),
    param_grid,
    cv=5,
    scoring='f1'
)

grid_search.fit(X_train, y_train)

print("Best Parameters:", grid_search.best_params_)
print("Best F1 Score (cross-validation):", grid_search.best_score_)

Best Parameters: {'C': 1, 'solver': 'liblinear'}
Best F1 Score (cross-validation): 0.7096508427699025


# Step 4: Extract new predictions from the best model

In [205]:
best_model = grid_search.best_estimator_

y_pred_tuned = best_model.predict(X_test)

from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score

accuracy_tuned = accuracy_score(y_test, y_pred_tuned)
precision_tuned = precision_score(y_test, y_pred_tuned)
recall_tuned = recall_score(y_test, y_pred_tuned)
f1_tuned = f1_score(y_test, y_pred_tuned)

print(f"Tuned Accuracy: {accuracy_tuned:.4f}")
print(f"Tuned Precision: {precision_tuned:.4f}")
print(f"Tuned Recall: {recall_tuned:.4f}")
print(f"Tuned F1-score: {f1_tuned:.4f}")

Tuned Accuracy: 0.7821
Tuned Precision: 0.7612
Tuned Recall: 0.6892
Tuned F1-score: 0.7234


# Step 5: Before/After Comparison Table

First, calculate the matrix for the original model (from Task 3) to calculate the comparison:

In [207]:
# Original model ki metrics (Task 3 wale model se)
accuracy_orig = accuracy_score(y_test, y_pred)
precision_orig = precision_score(y_test, y_pred)
recall_orig = recall_score(y_test, y_pred)
f1_orig = f1_score(y_test, y_pred)

# Comparison table banao
comparison = pd.DataFrame({
    'Metric': ['Accuracy', 'Precision', 'Recall', 'F1-score'],
    'Before Tuning': [accuracy_orig, precision_orig, recall_orig, f1_orig],
    'After Tuning': [accuracy_tuned, precision_tuned, recall_tuned, f1_tuned]
})

comparison

,Metric,Before Tuning,After Tuning
0,Accuracy,0.810056,0.782123
1,Precision,0.785714,0.761194
2,Recall,0.743243,0.689189
3,F1-score,0.763889,0.723404


## Hyperparameter Tuning Results

GridSearchCV tested 10 combinations (C values: 0.01, 0.1, 1, 10, 100, and
solvers: liblinear, lbfgs) with 5-fold cross-validation. The best parameters were
C=1 and solver='liblinear', with a cross-validation F1-score of 0.71.

Interestingly, the tuned model's performance on the original test set decreased slightly compared to the previous (default) model—accuracy dropped from 81.01% to 78.21%, and
F1-score dropped from 0.76 to 0.72. This is because GridSearchCV selects its best model based on cross-validation performance, which doesn't always match the results on the final test set—this is a normal variation when the dataset is small. This finding is also valuable: it teaches that tuning doesn't always guarantee better results, and in the real world, it's always important to validate the model on an independent test set.

### Before vs After Tuning

| metric | Before Tuning | After Tuning |
|-----------|----------------|-------------|
| Accuracy | 0.8101 | 0.7821 |
| Precision | 0.7857 | 0.7612 |
| Recall 0.7432 | 0.6892 |
| F1-score | 0.7639 | 0.7234 |

### Best Hyperparameters Found
- C: 1
- Solver: liblinear
- Cross-validation F1-score: 0.7097

**Note:** Tuning improved the cross-validation F1-score, but performance on the final test set decreased slightly—this shows that the results of hyperparameter tuning depend on the size and split of the dataset, and improvement is not always guaranteed.